### PR TITLE
ciao-launcher: Add Config package support for  ciao-launcher

### PR DESCRIPTION
**Pull request by obedmr**
Add config package support for enabling Scheduler-provided cluster
configuration.

ciao-launcher start commands should look like:
```
$ ./ciao-launcher --cacert=/path/to/cacert --cert=/path/to/cert \
       --server=<server-ip> --network=<cn|nn>
```

The options that were removed are:
```
--compute-net
--mgmt-net
--disk-limit
--mem-limit
```

Implements part of issue #66

Above configurations are provided by Scheduler at launch stage.

Signed-off-by: Munoz, Obed N <obed.n.munoz@intel.com>

### CODE REVIEW

**Comment by markdryan** [comment]
@obedmr Could you update the Readme.md (https://github.com/01org/ciao/tree/master/ciao-launcher) as the command line options have changed?

File: ciao-launcher/main.go [hunk 1]
```diff
@@ -333,6 +331,22 @@ func connectToServer(doneCh chan struct{}, statusCh chan struct{}) {
 		dialCh <- err
 	}()
 
+	var clusterConfig payloads.Configure
+	var data []byte
+	for {
```
Comment (markdryan) [inline ciao-launcher/main.go]: I don't really like this loop.  It will use loads of CPU and might never exit.  It will also stop launcher from exiting cleanly when there's a SIGTERM.  There must be a better way to access the SSNTP provisioned cluster config.  Is this data guaranteed to be present once dial suceeds?  If so we just need to move this call down a little bit, 
```
		select {
		case err := <-dialCh:
			dialing = false
			if err != nil {
				break DONE
			}

                        < Could we put the call to client.ClusterConfiguration() here?>
```

File: ciao-launcher/main.go [hunk 2]
```diff
@@ -97,12 +99,8 @@ func init() {
 	flag.StringVar(&serverURL, "server", "", "URL of SSNTP server")
 	flag.StringVar(&serverCertPath, "cacert", "/etc/pki/ciao/CAcert-server-localhost.pem", "Client certificate")
 	flag.StringVar(&clientCertPath, "cert", "/etc/pki/ciao/cert-client-localhost.pem", "CA certificate")
-	flag.StringVar(&computeNet, "compute-net", "", "Compute Subnet")
-	flag.StringVar(&mgmtNet, "mgmt-net", "", "Management Subnet")
 	flag.Var(&networking, "network", "Can be none, cn (compute node) or nn (network node)")
 	flag.BoolVar(&hardReset, "hard-reset", false, "Kill and delete all instances, reset networking and exit")
-	flag.BoolVar(&diskLimit, "disk-limit", true, "Use disk usage limits")
-	flag.BoolVar(&memLimit, "mem-limit", true, "Use memory usage limits")
```
Comment (markdryan) [inline ciao-launcher/main.go]: When you remove this line the defaults for these globals will change from true to false, i.e., you'll disable mem and disk checking by default, which is not what we want.

File: ciao-launcher/main.go [hunk 3]
```diff
@@ -528,6 +542,14 @@ func startLauncher() int {
 	timeoutCh := make(chan struct{})
 	signal.Notify(signalCh, syscall.SIGINT, syscall.SIGTERM, syscall.SIGHUP)
 
+	configCh := make(chan payloads.Configure)
+	go connectToServer(doneCh, statusCh, configCh)
+	clusterConfig := <-configCh
```
Comment (markdryan) [inline ciao-launcher/main.go]: We might as well move the networking initialisation code into connectToServer after Dial has returned.  This way we can get rid of the channel.

File: ciao-launcher/main.go (hunk 1 above)
Comment (markdryan) [inline ciao-launcher/main.go]: Actually, I think it will have to work like this.  Otherwise I don't see how we can ensure that we have the correct configuration data before we start processing SSNTP commands.  Your original code obviously does ensure this, but not in quite the right way.

File: ciao-launcher/main.go [hunk 4]
```diff
@@ -333,6 +331,22 @@ func connectToServer(doneCh chan struct{}, statusCh chan struct{}) {
...
 
+	var clusterConfig payloads.Configure
+	var data []byte
+	for {
+		data = client.ClusterConfiguration()
+		if data != nil {
+			break
+		}
+	}
+	err := yaml.Unmarshal(data, &clusterConfig)
```
Comment (sameo) [inline ciao-launcher/main.go]: I just pushed commit 7d9337799a7442190a0b635da0ea3aa16298382d that changes the client.ClusterConfiguration() call to return a payloads.Configure structure instead.

File: ciao-launcher/main.go (hunk 2 above)
Comment (mrkz) [inline ciao-launcher/main.go]: @markdryan so then `disk-limit` and `mem-limit` should default to **true**?
We had a conversation with @sameo and agreed to have them set to false by default, but of course those can be changed if this is not correct.

Comment (markdryan) [inline ciao-launcher/main.go]: Yep I think they should be true by default.  Otherwise launcher will not generate snntp.FULL messages and a node could become overwhelmed.  Even if we were to switch the defaults to false, we should do this in a separate commit so it's obvious that a change has been made.

File: ciao-launcher/main.go (hunk 3 above)
Comment (obedmr) [inline ciao-launcher/main.go]: hi @markdryan,  I'm still having trouble with this, as far I can see, it could be good if we only have network variables (computeNet and mgmtNet), but, in CluserConfiguration we also have the diskLimit and memLimit variables which are used by the overseer.go. So, if it is loaded in a go routine, not sure how to pass those variables to the overseer which is consuming the main.go context. Do you have a recommendation for this?,  I updated my commit by applying the rest of comments.

Comment (markdryan) [inline ciao-launcher/main.go]: @obedmr This is a good point.  There'll be a race condition if we don't also delay the start of the overseer until after Dial has completed and we've retrieved our ClusterConfiguration and initialised the 4 global variables.  I think this should work, although I haven't tried.

Comment (obedmr) [inline ciao-launcher/main.go]: @markdryan @sameo, I updated my commit, when you have a chance could you please take a look on it and comment.

Comment (markdryan) [inline ciao-launcher/main.go]: @obedmr I'll take a look this afternoon.

**Comment by markdryan** [comment]
The unit tests no longer compile with this patch applied.

ClusterConfiguration needs to be added to instanceTestState in instance_test.go

**Comment by obedmr** [comment]
thanks @markdryan, I will update unit tests and see how we can start overseer after dial is succeed

File: ciao-launcher/instance_test.go [hunk 5]
```diff
@@ -250,6 +250,10 @@ func (v *instanceTestState) deleteInstance(t *testing.T, ovsCh chan interface{},
 	}
 }
 
+func (v *instanceTestState) ClusterConfiguration() (payloads.Configure, error) {
+	return payloads.Configure{},nil
+}
+
```
Comment (markdryan) [inline ciao-launcher/instance_test.go]: This file needs to be gofmted.  The formatting is incorrect.

**Comment by markdryan** [comment]
gofmt needs to be run on both the launcher files.

File: ciao-launcher/main.go [hunk 6]
```diff
@@ -356,6 +353,28 @@ DONE:
 			if err != nil {
 				break DONE
 			}
+			clusterConfig,err := client.conn.ClusterConfiguration()
+			if err != nil {
+				glog.Errorf("Unable to get Cluster Configuration %v", err)
+				dialCh <- err
```
Comment (markdryan) [inline ciao-launcher/main.go]: Is this a copy and paste error?  Maybe it was in the code I sent you.  I think this will deadlock.  There's no one to read from dialCh.  We can just remove this line.

File: ciao-launcher/main.go [hunk 7]
```diff
@@ -356,6 +353,28 @@ DONE:
...
+				ch := initNetworking(ctx)
+				err := <-ch
+				if err != nil {
+					glog.Errorf("Failed to init network: %v\n", err)
+					break DONE
+				}
+			}
+			defer shutdownNetwork()
+
+			ovsCh = startOverseer(&wg, client)
```
Comment (markdryan) [inline ciao-launcher/main.go]: connectToServer is now too complex

```
16 main connectToServer ciao-launcher/main.go:320:1
```

We'll need to split it up.

**Comment by markdryan** [comment]
Make sure you add the copyright notice to all the new files.

**Comment by markdryan** [comment]
@obedmr Apart from the 3 comments (gofmt, gocyclo and the deadlock) the launcher change looks fine.  I'm still a bit confused about whether the defaults for memlimit and disklimit will change.  This PR should not change the defaults.  I can't quite figure out what would happen if the current PR was merged.

**Comment by markdryan** [comment]
One more comment.  What is the plan for the cnci-agent, which also has mgmt-net and compute net parameters?

**Comment by tpepper** [comment]
On Fri 03 Jun at 08:24:25 -0700 notifications@github.com said:
> Make sure you add the copyright notice to all the new files.

See "template.go" in the toplevel directory.

-- 
Tim Pepper -- Linux OS Engineering
Intel Open Source Technology Center

File: ciao-launcher/main.go (hunk 7 above)
Comment (obedmr) [inline ciao-launcher/main.go]: @markdryan, not sure what happened by I'm getting 14, is that ok, or should we still split it?
14 main connectToServer ciao-launcher/main.go:316:1

**Comment by obedmr** [comment]
@markdryan @sameo, I'm just waiting for @mrkz pull request #149 to be updated and then I'll do more testing. For now, I'm taking a look on (memLimit, diskLimit) and cnci-agent things.

File: ciao-launcher/main.go (hunk 7 above)
Comment (markdryan) [inline ciao-launcher/main.go]: 14 is fine but I'm getting 16.  Have you rebased your patch set against the latest version of master?  If you have and you are still getting 14, perhaps I merged your patch incorrectly to my local workspace.  In any case, if there really is a problem here, Travis CI will fail the build when you update your PR.

**Comment by obedmr** [comment]
@markdryan, now ``connectToServer`` function gives 15 with gocyclo, I split the network initialization in another function,  regarding the memLimit and diskLimit, they're being initialized at https://github.com/01org/ciao/blob/master/configuration/configuration.go#L50-57

In case that they're not configured in yaml file, they will have ``true`` by default.

File: ciao-launcher/main.go [hunk 8]
```diff
@@ -318,6 +315,16 @@ func processCommand(conn serverConn, cmd *cmdWrapper, ovsCh chan<- interface{})
 	}
 }
 
+func startNetwork() error {
+	if networking.Enabled() {
```
Comment (markdryan) [inline ciao-launcher/main.go]: This code is not quite right.  The original networking code also selected on the signal channel so that the code that inits the network could not block a launcher exit via CTRL-C or SIGTERM.  This code has been deleted.  Launcher will still be able to shutdown but it will no longer be able to shutdown cleanly.

You don't have the signal channel in the connectToServer go routine but you do have doneCh which is closed when we received a signal

```
func startNetwork(doneCh chan struct{}) error {
	  if networking.Enabled() {
                ctx, _ := context.WithCancel(context.Background())
		ch := initNetworking(ctx)
		select {
		case <-doneCh:
			glog.Info("Received terminating signal.  Quitting")
			cancelFunc()
			return fmt.Error("Shutting down")
		case err := <-ch:
			if err != nil {
				glog.Errorf("Failed to init network: %v\n", err)
				return err
			}
		}
}
```

**Comment by markdryan** [comment]
@obedmr Looks okay but I'd like to give it a quick test.  Will do so first thing tomorrow.  Is there anything preventing us from merging this now?  If not and it works, I'll merge tomorrow morning.

**Comment by obedmr** [comment]
@mrkz, cool. I will be testing this Today in a clean environment an will
let you know if there's something else.

File: ciao-launcher/main.go [hunk 9]
```diff
@@ -369,6 +385,23 @@ DONE:
 			if err != nil {
 				break DONE
 			}
+			clusterConfig, err := client.conn.ClusterConfiguration()
+			if err != nil {
+				glog.Errorf("Unable to get Cluster Configuration %v", err)
```
Comment (markdryan) [inline ciao-launcher/main.go]: @obedmr  Sorry, about this but I just noticed that we need to call

client.conn.Close()

here to shutdown the connection to the SSNTP server.

File: ciao-launcher/main.go [hunk 10]
```diff
@@ -369,6 +385,23 @@ DONE:
...
+				break DONE
+			}
+			computeNet = clusterConfig.Configure.Launcher.ComputeNetwork
+			mgmtNet = clusterConfig.Configure.Launcher.ManagementNetwork
+			diskLimit = clusterConfig.Configure.Launcher.DiskLimit
+			memLimit = clusterConfig.Configure.Launcher.MemoryLimit
+			err = startNetwork(doneCh)
+			if err != nil {
+				glog.Errorf("Failed to start network: %v\n", err)
```
Comment (markdryan) [inline ciao-launcher/main.go]: We need to call client.conn.Close() here as well.

Comment (markdryan) [inline ciao-launcher/main.go]: I think it would also be useful to add a log statement that outputs the values of all the configuration variables to the launcher logs.

File: ciao-launcher/main.go (hunk 9 above)
Comment (obedmr) [inline ciao-launcher/main.go]: Done

**Comment by obedmr** [comment]
@markdryan missing client.conn.Close() function calls were added and printClusterConfig function was added
